### PR TITLE
Create .gitattributes file for auto-formatting of files

### PR DESCRIPTION
**Pull request by awemany**
To improve upon the .formatted-files logic. See commit message for details.

### CODE REVIEW

File: src/Makefile.am
Comment (ftrader) [inline src/Makefile.am]: This seems to be assuming that all files in .gitattributes shall be subjected to bitcoin-clang-format filter.

I would suggest filtering for those files explicitly when constructing the list to pass.

Comment (awemany) [reply src/Makefile.am]: Excellent point. Will rework. Is a grep in the pipe good enough here?

EDIT: Added grep.

Comment (ftrader) [reply src/Makefile.am]: Seems alright as a solution.

File: doc/developer-notes.md
Comment (gandrewstone) [inline doc/developer-notes.md]: trailing-comment.py needs to run before clang-format.  But best of all, call contrib/devtools/clang-format.py.  Using this script will ensure that your code always matches make check-formatting...

**Comment by gandrewstone** [comment]
I dislike that a nice, self-documenting filename like .formatted-files becomes .gitattributes.  Especially for a cool but very obscure feature that will likely be used by few people.  Is there any way to keep the original name?  Symlink?

**Comment by ftrader** [comment]
How about a general rule in the .gitattributes for all .h / .cpp files under src/ , that call a script which performs the action only if the file is listed in .formatted-files ?

Long-term we should be getting to where we process all C++ source files anyway...

**Comment by awemany** [comment]
@thezerg: A symlink should be possible, yes. Probably with some explanation at the top of gitattributes (I think #-comments work there) that if you reach there from .formatted-files, the reason why it looks so funny now is explained. I have added that now.

@ftrader: 
> What about a general rule in the .gitattributes for all .h / .cpp files under src/ , that call a script which performs the action only if the file is listed in .formatted-files ?

Hmm, but that appears to me to duplicate the effort of walking through lists of files. Why another indirection here?

**Comment by ftrader** [comment]
I would be opposed to symlinks. They should be avoided within the project sources - they are not a very portable feature, and to the extent that our build uses any (I did find some use in the OSX build path and 3rd party lib Makefiles), I think it's regrettable.

>Hmm, but that appears to me to duplicate the effort of walking through lists of files. Why another indirection here?

Two reasons:
- to preserve the neat filenaming that thezerg likes
- to avoid using the VCS-specific file for this as much as possible. The less specific info we need in the .gitattributes, the better it would be I feel. Might make it easier to hook the checks into other systems.

File: doc/developer-notes.md
Comment (awemany) [reply doc/developer-notes.md]: Ok, I have modified the PR to do so. This needed a change to the clang-format.py script, which I have now added as a separate commit.

**Comment by awemany** [comment]
> I would be opposed to symlinks. They should be avoided within the project sources - they are not a very portable feature, and to the extent that our build uses any (I did find some use in the OSX build path and 3rd party lib Makefiles), I think it's regrettable.

Ok. So if it isn't .gitattributes, should I add a rule to the Makefile that auto-creates a .gitattributes with the right content?

The problem here would be that this .gitattributes might also grow to contain other info and then that starts to get annoying.


>    to avoid using the VCS-specific file for this as much as possible. The less specific info we need in the .gitattributes, the better it would be I feel. Might make it easier to hook the checks into other systems.

Ok that's a fair point, though I am personally not worried. I'll wait for @thezerg for the final decision on this.

**Comment by gandrewstone** [comment]
I like ftrader's idea of wildcarding in .gitattributes and having a script in devtools that checks .formatted-files for the file.  That idea has 2 additional advantages: it gives people a "hook" to add other check-in tests (without having to know anything git specific), and lets a person run the check-in tests manually before actually checking in.

**Comment by gandrewstone** [comment]
changing to WIP for my own tracking, let me know when you make changes...

**Comment by awemany** [comment]
I have updated the pull request. It now checks the list in .formatted-files and otherwise leaves that file alone. As git has to do that repeatedly, a minor downside here is that the commands that deal with auto-formatting become noticeably slower.

**Comment by awemany** [comment]
The failing test here is proxy_test.py, which appears to be completely unrelated to my code.

**Comment by sickpig** [comment]
@awemany 
rekick travis job.

**Comment by sickpig** [comment]
@awemany 
you are good to go now :smile: 
sometimes we have some spurious failure on travis (this seems to be something we share with a lot of other projects using it).

**Comment by sickpig** [comment]
@awemany do you think this is still WIP or that are ready to be merged? (fwiw is ready to be merged imho)

**Comment by awemany** [comment]
@sickpig: I am still moderately unhappy with the reduced performance now (and want to look into improving that at some point) but this still will make it a lot quicker and easier for me to fulfill the formatting requirements. So yes, I also think it is ready to be merged.

**Review by sickpig — APPROVED** [review]

**Comment by sickpig** [comment]
approved

**Comment by sickpig** [comment]
@gandrewstone I think this ready to merge and a really needed change.

**Comment by gandrewstone** [comment]
Great, glad to see this come in! Please kick me on slack if an older PR is taken out of WIP.

**Pull request merged**
